### ISSUE TITLE
[RFC] Verl recipe for image generation model

### ISSUE THREAD

**Issue by Franklin-Zhang0** [verl-project/verl#2136]
# Motivation
Reinforcement learning for autoregressive image generation with textual chain-of-though is attracting growing interest. A recent project, [ReasonGen-R1](https://github.com/Franklin-Zhang0/ReasonGen-R1), employs GRPO via Verl to enhance an image generation model’s ability to follow instructions. We plan to contribute this training pipeline as a Verl recipe to benefit the wider community.

# Overall Structure
```
/recipe/image_generation
      config/
           image_generation_rl.yaml
           image_generation_sft.yaml
      main.py
      ray_trainer.py
      sft_trainer.py
      fsdp_worker.py
      dp_actor.py
      hf_rollout.py
      datasets/
            rl_datasets.py
            sft_datasets.py
      Janus/
```

# Proposed Major Changes
## New Function or Classes
`ImageRewardModelWorker` in fsdp_worker.py: We need to implement a reward model that takes in the generated images and prompts to assess the image quality and instruction following. The overall structure is similar to the current RewardModelWorker.
`Janus-model`: An example model class for RL training. Need to implement inside the recipe for interleaved image-text generation and official genrate and forward function is not released by deepseek.
`AdaptiveEntropyCoefficient` in `dp_actor.py`: Adaptive entorpy loss coefficient for stable training in text-image interleaved RL training. It updates using the target entropy and the entropy of output logits.

## Function or Classes that needs modification:
`FSDPSFTTrainer` in FSDPSFTTrainer: support sft training for image generation model
`HFSFTDataset` in `sft_dataset.py` and `RLDataset` in `rl_dataset.py`: support data loading and formating for image generation
`_build_model_optimizer` in `ActorRolloutRefWorker`: modify to support janus loading
`update_actor` in `DataParallelPPOActor`: handle the adaptive entropy and seperate entropy computation for text and image 
`hf_rollout`: return generated images and generated texts as seperate output.
`fit` in ray_trainer.py: support group_filtering in dapo

# CC
@eric-haibin-lin

**Comment by vermouth1992** [comment]
Could you give an example of image generation model?

**Comment by Franklin-Zhang0** [comment]
An example image generation model we want to integrate is Janus-Pro, which is also the base model in [ReasonGen-R1](https://github.com/Franklin-Zhang0/ReasonGen-R1)

**Comment by eric-haibin-lin** [comment]
Thanks! What code goes in Janus-Pro? Is it possible to reuse from huggingface/transformers instead of copying all its code?

**Comment by Franklin-Zhang0** [comment]
I just checked the latest huggingface/transformers and found that they updated the janus-pro model last week. 
So, yes. We can reuse code from there. For pure image generation or pure text generation, reusing their code would be enough.

We might still need to customize some generation functions for generating interleaved text and image, which is not originally supported by janus-pro.

**Comment by eric-haibin-lin** [comment]
Thanks. For rollout is only hf rollout supported? What's the issue with vllm/sglang?

**Comment by Franklin-Zhang0** [comment]
By the time I implement it, janus is not supported by vllm/sglang. I think sglang support janus now. However, I think it still doesn't support interleaved generation with text and images. But maybe I can do some customization to integrate vllm/sglang.

**Comment by eric-haibin-lin** [comment]
I'm okay with proceeding with the current plan, and having hf rollout as the initial PR to land this recipe. sglang/vllm rollout can be added as separate enhance PRs.
cc @zhaochenyang20 for sglang feature visibility

**Comment by zhaochenyang20** [comment]
> By the time I implement it, janus is not supported by vllm/sglang. I think sglang support janus now. However, I think it still doesn't support interleaved generation with text and images. But maybe I can do some customization to integrate vllm/sglang.

SGLang support Janus. What do you need us to support? I can ask @mickqian

**Comment by mickqian** [comment]
FYI, image-generation capabilities are not supported in sglang, nor are they in sglang's roadmap. Not sure if it  conflicts with sglang's goal.

**Comment by pimang62** [comment]
Is there any updated feature now? I'm hopefully waiting for RL training in image generation task. 🙏

**Comment by Kirrito-k423** [comment]
定时清理检查发现：该 issue 最近一次非当前运行账号（Kirrito-k423）的公开回复或公开引用是 2025-09-29T11:30:37Z（actor=pimang62, type=commented），截至 2026-07-09 已超过 6 个月。按照社区 issue 定时清理标准，本次先关闭该 issue。

Scheduled maintenance check: the latest public reply or public reference from an account other than the current runner (Kirrito-k423) was at 2025-09-29T11:30:37Z (actor=pimang62, type=commented), which is more than 6 months before 2026-07-09. Under the scheduled issue cleanup rule, this issue is being closed for now.

证据 / Evidence: latest_non_self_reply_or_reference_at=2025-09-29T11:30:37Z; comments_checked=10; timeline_public_events_checked=15; issue=https://github.com/verl-project/verl/issues/2136; timeline API=https://api.github.com/repos/verl-project/verl/issues/2136/timeline.

如果作者或社区成员仍能提供新的复现、证据或需求背景，欢迎继续评论请求重开，或创建新的 issue 并引用这里的上下文。

If the author or community can still provide a fresh reproduction, evidence, or requirement context, please comment to request reopening or open a new issue and link back to this context.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
